lib/puppet/type/nssdb.rb

```diff
@@ -0,0 +1,53 @@
+Puppet::Type.newtype(:nssdb) do
+  desc "Generates an empty NSS database"
+
+  ensurable
+
+  newparam(:nssdb_dir, :namevar => true) do
+    desc "Path to NSS database directory"
+    isrequired
+  end
+
+  newparam(:password_file) do
+    desc "Path to file containing the NSS database password"
+    isrequired
+  end
+
+  newparam(:owner, parent: Puppet::Type::File::Owner) do
+    desc "Specifies the owner of the NSS database directory and files. Valid options: a string containing a username or integer containing a uid."
+  end
+
+  newparam(:group, parent: Puppet::Type::File::Group) do
+    desc "Specifies a permissions group for the NSS database directory and files. Valid options: a string containing a group name or integer containing a gid."
+  end
+
+  newparam(:mode, parent: Puppet::Type::File::Mode) do
+    desc "Specifies the permissions mode of the NSS database files. Valid options: a string containing a permission mode value in octal notation."
+  end
+
+  autorequire(:file) do
+    [self[:password_file]]
+  end
+
+  def generate
+    file_opts = {}
+
+    [:owner, :group, :mode].each do |param|
+      file_opts[param] = self[param] unless self[param].nil?
+    end
+
+    excluded_metaparams = [:before, :notify, :require, :subscribe, :tag]
+
+    Puppet::Type.metaparams.each do |metaparam|
+      unless self[metaparam].nil? || excluded_metaparams.include?(metaparam)
+        file_opts[metaparam] = self[metaparam]
+      end
+    end
+
+    file_opts.merge({ :path => self[:nssdb_dir] })
+    file_opts.merge({ :ensure => self[:ensure] == :absent ? :absent : :directory })
+    file_opts.merge({ :recurse => true })
```

Comment:
`merge` is not an inplace update to the hash. Ruby tends to have functions with `!` at the end, such as `merge!` to differentiate "destructive" actions. That being said, I think you could just define these at the top like we do in other resources:

https://github.com/theforeman/puppet-certs/blob/master/lib/puppet/type/keystore.rb#L33-L36

Reply:
Thanks @ehelms !

Originally I was doing something like

```ruby
resources = []
resources.append(Puppet::Type.type(:file).new(common_file_opts.merge({ :new_opt_key => :new_opt_value })))
resources.append(Puppet::Type.type(:file).new(common_file_opts.merge({ :new_opt_key => :new_opt_different_value })))
```

And I didn't change the pattern when I moved to multiple new option keys (at this point it would have been broken without `merge!`), and then moved away from managing multiple resources entirely.

Now I moved all of these static attributes to the top to avoid using `merge` or `merge!` entirely, since it's no longer needed.